src/types/did_parser.rs

```diff
@@ -29,53 +29,24 @@ peg::parser! {
         ///        }
         ///   });
         /// ```
-        pub rule ethr_did() -> MethodAndId
-            = method:method() id:id() { MethodAndId { method, id } }
+        pub rule ethr_did() -> Did
+            = method:method() ":" network:network()? account:account() { Did { method, network: network.unwrap_or_default(), account } }
 
         rule method() -> Method
-            = "ethr:" { Method::Ethr } / expected!("the only supported method is `ethr`")
+            = "ethr" { Method::Ethr } / expected!("the only supported method is `ethr`")
 
-        rule id() -> Id
-            = id: id_no_network() { id } / id:id_and_network() { id }
+        rule network() -> Network
+            // chain id networks
+            = "0" i("x") digits:$(hex_digit()+) ":" { Network::from(digits) }
+            // named networks
+            / "mainnet:" { Network::Mainnet }
+            / "sepolia:" { Network::Sepolia }
+            / "goerli:" { Network::Goerli }
+            / expected!("the only supported networks are `mainnet`, `goerli`, `sepolia`, and chain id")
 
-        rule id_and_network() -> Id
-            = network:ethr_network() ":" key:address_or_hex() { Id { chain: network, public_key: key } }
-
-        // the default chain is Mainnet, if it's ommitted we default to Mainnet
-        rule id_no_network() -> Id
-            = key:address_or_hex() { Id { chain: Default::default(), public_key: key } }
-
-        rule ethr_network() -> ChainId
-            = network_chain_id:network_chain_id() { network_chain_id }
-            / network_name:(mainnet() / goerli()) { network_name }
-
-        rule address_or_hex() -> AddressOrHexKey
-            = address:ethereum_address() { address }
-            / key:public_key_hex() { key }
-
-        rule mainnet() -> ChainId
-            = i("mainnet") { ChainId::Mainnet }
-
-        rule goerli() -> ChainId
-            = i("goerli") {
-                #[allow(deprecated)]
-                ChainId::Goerli
-            }
-
-        rule network_chain_id() -> ChainId
-            = "0" i("x") digits:$(hex_digit()+) {
-                ChainId::from(digits)
-            }
-
-        rule ethereum_address() -> AddressOrHexKey
-            = "0" i("x") digits:$(hex_digit()*<40>) {
-                AddressOrHexKey::Address(Address::from_slice(&hex::decode(digits).unwrap()))
-            }
-
-        rule public_key_hex() -> AddressOrHexKey
-            = "0" i("x") digits:$(hex_digit()*<66>) {
-                AddressOrHexKey::HexKey(hex::decode(digits).unwrap())
-            }
+        rule account() -> Account
```

Comment:
The choice is an address or a public key.   An address is 40 characters, a key 60.   We don't need to enforce limitations on those values but it would be correct to parse them into the proper context.

Reply:
We should also remain consistent, in `resolve_did` method, we accept a `public_key` argument, but as of now it's just and Ethereum Address, which can become confusing

Not completely related, but do you know how the public key works with the registry? afaiu, the registry works with ethereum addresses, and I could not find anywhere that describes resolution based on a public key

Reply:
yeah, I take some responsibility for this because I think I frequently use these two terms ambidextrously.   I hope I haven't made it confusing.    When I say public key I often mean address.

The address should be H(PK)[20..].   That is take the keccak256 of the public key, then the last 20 bytes are the address.   We should be able to prove that by using ethkey to generate a few random keys and make sure it works.   I didn't want to post my PK to avoid name checking it here. 

Also an address is usually noted by prepending 0x as a hex address.    A public key is just numbers so just data and no prefix.

From the standpoint of this change I would not include this support for querying PKs in this PR.   It is fine if that is a TODO.   The parser should know when a PK vs and ADDY is encountered.   Like I said it should parse out the context correctly if it ever encounters that case.  However right now resolving by address is sufficiently useful to get us started.    I don't know of any use cases in the wild where people are calling into things by their PK.   I doubt many people even know their PK so we can add an issue for this and leave it for a rainy day.

Reply:
Okay, would make sense to follow up or capture that in a ticket.

Reply:
added the ticket [here](https://github.com/xmtp/didethresolver/issues/34).